core/domain/stats_jobs_one_off.py

```diff
@@ -1280,29 +1291,63 @@ def map(exp):
                 exp.id, version)
             exp_at_version = exp_models.ExplorationModel.get_version(
                 exp.id, version)
-            # Delete all old models.
-            if all_models[version - 1] is None:
-                if commit_log.commit_type != 'revert':
-                    yield ('Missing model without revert commit', exp.id)
-                    return
-
-                # Is a revert commit.
-                revert_to_version = commit_log.commit_cmds[0]['version_number']
+
+            if commit_log is None:
+                prev_exp = exp_models.ExplorationModel.get_version(
+                    exp.id, version - 1)
+                old_states = prev_exp.states
+                new_states = exp_at_version.states
+                inferred_change_list = []
+                for old_state in old_states:
+                    if old_state not in new_states:
+                        inferred_change_list.append(
+                            exp_domain.ExplorationChange({
+                                'cmd': exp_domain.CMD_DELETE_STATE,
+                                'state_name': old_state
+                            }))
+
+                for new_state in new_states:
+                    if new_state not in old_states:
+                        inferred_change_list.append(
+                            exp_domain.ExplorationChange({
+                                'cmd': exp_domain.CMD_ADD_STATE,
+                                'state_name': new_state
+                            }))
+                exp_versions_diff = exp_domain.ExplorationVersionsDiff(
+                    inferred_change_list)
+
+                if all_models[version - 1] is not None:
+                    all_models[version - 1].delete()
                 new_exp_stats_dicts.append(
                     stats_services.get_stats_for_new_exp_version(
-                        exp.id, version, exp_at_version.states, None,
-                        revert_to_version).to_dict())
+                        exp.id, version, exp_at_version.states,
+                        exp_versions_diff, None).to_dict())
             else:
-                all_models[version - 1].delete()
                 change_list = (
                     [exp_domain.ExplorationChange(commit_cmd)
                      for commit_cmd in commit_log.commit_cmds])
                 exp_versions_diff = exp_domain.ExplorationVersionsDiff(
                     change_list)
-                new_exp_stats_dicts.append(
-                    stats_services.get_stats_for_new_exp_version(
-                        exp.id, version, exp_at_version.states,
-                        exp_versions_diff, None).to_dict())
+
+                # Delete all old models.
+                if all_models[version - 1] is None:
+                    if commit_log.commit_type != 'revert':
+                        yield ('Missing model without revert commit', exp.id)
```

Comment:
You probably want to pass along more identifying info here?

Reply:
I am not sure. I checked out the exps which failed and none of them had a revert commit. It seems to be happening due to some other reason. I'll take a look through the code and then decide what to add here.